[Ruby] new concept exercise - `nil` (#1953)
exercism/v3: languages/ruby/exercises/concept/booleans/.docs/introduction.md

## True, False, and Nil

> True and false logical states are represented with `true` and `false` in Ruby. These may either be used or returned as literals on their own, or as a result of logical or comparison methods.

Author: ```suggestion
True and false logical states are represented with `true` and `false` in Ruby. These may either be used as literals on their own, or as a result of logical or comparison methods.
```

```ruby
true_variable = true
false_variable = false

true && false

> # => false is returned as a result of the logical method

Reviewer: ```suggestion
# => false
```

1 < 2

> # => true is returned as a result of the comparison method

Reviewer: ```suggestion
# => true
```

```

[Nil][nil-dictionary] is an English word meaning "nothing" or "zero". In Ruby, `nil` is a value used to express the _absence_ of an object.

```ruby
# I do not have a favorite color
favorite_color = nil
```

> In other programming languages, `null` or `none` values might play a similar role.

Reviewer 2: ```suggestion
In other programming languages, `null` or `none` values may play a similar role.
```

## _Truthy_ and _falsey_

> In Ruby, all objects evaluate to a _truthy_ or _falsey_ value when they are encountered in a boolean context (like an `if` expression). All data is considered _truthy_ **except** for `false` and `nil`.

Reviewer: ```suggestion
When evaluating objects in `if` statements or other boolean contexts, all objects evaluate to a _truthy_ **except** for `false` and `nil`.
```

Reviewer 2: ```suggestion
When evaluating objects in `if` statements or other boolean contexts, all objects evaluate as _truthy_ **except** for `false` and `nil`.
```

> In Ruby, all objects evaluate to a _truthy_ or _falsey_ value when they are encountered in a boolean context (like an `if` expression). All data is considered _truthy_ **except** for `false` and `nil`.

Reviewer 2: ```suggestion
In Ruby, all objects evaluate to a _truthy_ or _falsey_ value when they are encountered in a boolean context (conditional statement). All values and objects evaluated as _true_ and considered _truthy_ **except** for `false` and `nil`, referred to as _falsey_.
```

I still don't like my changes all that much, except for the conditional statement.

If statements should not produce a true or false value, if they do, it is often an indication that it should be a conditional statement instead.  The "truthy/falsey" comes from the if statement only because everything that is an object has this attribute, and the if statement will always return something when evaluated/invoked.

Author: ```suggestion
In Ruby, all objects evaluate to a _truthy_ or _falsey_ value when they are encountered in a boolean context (conditional statement). All values and objects are considered _truthy_. Only `false` and `nil` are considered _falsey_ values.
```

maybe?

Reviewer 2: All objects evaluate as _truthy_ or _falsey_. Not as a value, as the idea is there, but not the "truthy" value, _per se_. I would drop the word "value" at the end of the last sentence as well.

Reviewer 2: Have we introduce the difference in definition of "boolean" and "Boolean" as we use it?  The "boolean" to indicate that we aren't necessarily looking for the formal `true` or `false` objects, but the _truthy_ or _falsey_ evaluation?

Author: would `logical context` be more clear?

Reviewer 2: As opposed to "conditional statement" or "boolean" or…

## Control flow

> _Truthy_ and _falsey_ values are useful in the context of control flow. Like in procedural languages, Ruby has an _if...else_ construct, but it is often more common to use _if_ as a modifier to control the evaluation of an expression.

Reviewer 2: ```suggestion
_Truthy_ and _falsey_ evaluations are useful in the context of control flow. Like in procedural languages, Ruby has an `if`...`else` construct.  It is common to use `if` as "guarding" statement.
```

Still needs some work as well,  I think.  (Writing is hard.)

Author: ```suggestion
_Truthy_ and _falsey_ evaluations are useful in the context of control flow. Like in procedural languages, Ruby has an `if`...`else` construct, but it is often more common to use `if` as a "guarding" statement to modify the evaluation of an expression.
```

I wanted to have the word modify in here because that is [how they are referred to in the docs](https://docs.ruby-lang.org/en/2.0.0/syntax/control_expressions_rdoc.html#label-Modifier+if+and+unless)

Reviewer 2: We can always patch the docs. ;)

`but it is often more common` to `it may be more common` (I have not measured, and so don't know, and it may be in one field and not another.)

Author: ```suggestion
_Truthy_ and _falsey_ evaluations are useful in the context of control flow. Like in procedural languages, Ruby has an `if`...`else` construct, but it may be more common to use `if` as a "guarding" statement to modify the evaluation of an expression.
```

> ```ruby

Reviewer 2: ```ruby
def truthy
  'any object other than nil or false'
end

def falsey
  not truthy
end
```

Does this explain it in a way that makes sense?  The _falsey_ could also be expressed as:

```ruby
def falsey
  [nil, false][rand(2)]
end
```

And if the latter version of `falsey` exists it follows that we could say:

```ruby
def truthy
  not falsey
end
```

Author: Could it even be simplified to:

```ruby
def falsey
  nil || false
end
```

While it may always evaluate to false in this definition, it reads as "falsey returns nil or false" which may be easier to grok at an entry level.

Reviewer 2: Yes, or even reversed it would always return `nil`.  It think it is a good expression.

Author: ````suggestion
```ruby
def falsey
  nil || false
end

def truthy
  not falsey
end

````

if truthy
  # this block is evaluated
end

if falsey
  # this block is not evaluated
else
  # this block is evaluated
end

1 + 1 if truthy
# => this will evaluate and return 2

2 + 2 if falsey
# => the numbers are not added because of the modifier, nil is returned
```

> Sometimes `unless` is used in place of `if` to provide an inverse condition

Reviewer 2: If we prefer to always have a positive conditional statement, `unless` can be used to indicate the overall negativity of the statement.

Author: ```suggestion
While it is preferred to always use the positive form of a conditional statement, `unless` is provided when for an overall negative condition.
```

Reviewer 2: If it is preferred to always use the positive form…

Reviewer 2: ```suggestion
If it is preferred to use the positive form of a conditional statement, `unless` can provide for an overall negative statement.

    eat_desert unless too_full
```

Reviewer 2: I don't like the wording, still, what does `unless` communicate to us about the entire statement?  It is still a guard… the conditional statement is positive…

Author: ```suggestion
If it is preferred to use the positive form of a conditional statement. But rather than negating a negative statement, `unless` is provided as an alternative.  E.g. `eat_desert unless too_full`
```

Reviewer 2: I like that better.  Perhaps:

```suggestion
If it is preferred to use the positive form of a conditional statement. Rather than negating a negative statement, `unless` is provided as an alternative.  E.g. `eat_desert unless too_full`
```

```ruby
3 + 3 unless truthy
# => the numbers are not added because of the modifier, nil is returned

> 4 + 4 unless falsey

Reviewer 2: This would not be used, preferably.  We would prefer `if` with a positive conditional statement.

Author: I added a suggestion to the previous comment.  Do you think that is adequate? I think it would be useful for students to see this as it isn't uncommon (though undesired).
